### PR TITLE
Switch to stestr for test runner

### LINKED ISSUE #730: Switch to parallel test runner for python unit tests
### What is the expected enhancement?

We currently are using the default stdlib test runner for running our python unit tests. While this works and will execute the tests correctly it's quite limited in what it can do. The biggest limitations are that it can only serial execute tests and that it doesn't enable any results tracking. 

I propose that we switch the test runner to stestr: https://github.com/mtreinish/stestr which addresses these limitations and also provides users with an increased level of flexibility and configurability in how tests are run (admittedly I am a bit biased, since I do maintain the project).The list of features over the stdlib runner is quite long, but the most immediate advantage is that by switching we'll gain a significant speed up in test execution time by default.

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit switches the default test runner for qiskit-terra to use
stestr instead of the built-in runner from stdlib unittest. By default
this switches the default mode of execution to be parallel instead of
serial which can lead to significant speed ups in test execution time.

### Details and comments

Closes #730

### CODE REVIEW

**Comment by delapuente** [comment]
On hold, until we advance in the discussion at #730.

**Comment by mtreinish** [comment]
Just for comparison there is a noticeable speedup by running tests in parallel. Looking at the first python test run in travis it runs about ~1.9x faster:

From master (without this patch):
516.53 seconds https://travis-ci.org/Qiskit/qiskit-terra/jobs/436905020#L1325 

With this patch:
279.16 seconds https://travis-ci.org/Qiskit/qiskit-terra/jobs/436928335#L1327

**Comment by mtreinish** [comment]
It's even more noticeable on appveyor (because the nodes are slower and smaller)

With this patch:
73.2527 seconds https://ci.appveyor.com/project/QISKit/qiskit-sdk-py/build/1.0.2699/job/v05h7a2y1k3e5pxk#L849

On master (without this patch):
307.82 seconds https://ci.appveyor.com/project/QISKit/qiskit-sdk-py/build/1.0.2696/job/tsgv4rgw3tqlx8a0#L790

**Comment by diego-plan9** [comment]
> Just for comparison there is a noticeable speedup by running tests in parallel. Looking at the first python test run in travis it runs about ~1.9x faster:

> From master (without this patch):
> 516.53 seconds https://travis-ci.org/Qiskit/qiskit-terra/jobs/436905020#L1325

> With this patch:
> 279.16 seconds https://travis-ci.org/Qiskit/qiskit-terra/jobs/436928335#L1327

While on it - can you revise the comparisons, as the number look a bit off? Note that `master` executes the `online_tests`, which the PRs are not - and they are very likely a good source of time spent (and one that might be difficult to shave off, unfortunately).

**Comment by mtreinish** [comment]
@diego-plan9 ah good call I forgot about that. I chose the results from here and from #1085 for comparison. For travis:

unittest runner:
291.7 seconds https://travis-ci.org/Qiskit/qiskit-terra/jobs/440170798#L1445

stestr:
277.7 seconds https://travis-ci.org/Qiskit/qiskit-terra/jobs/439867792#L1316

which makes more sense to me than the earlier numbers, because the travis nodes are pretty fast so the gains by running in parallel with only fast executing tests should be less. It makes more of a difference on resource constrained environments like a laptop or appveyor. That being said for the online tests it should be a bigger difference because the slow online tests will eat a worker while the rest of the tests use the remaining 47 workers (we might need to tune the stestr scheduler to force that configuration for online tests though to get the best performance).

For appveyor:

unittest runner:
184.4 seconds https://ci.appveyor.com/project/QISKit/qiskit-sdk-py/builds/19434294/job/ehrwdm66e4414tti#L852

stestr:
101.3 seconds https://ci.appveyor.com/project/QISKit/qiskit-sdk-py/builds/19415964/job/jhruacgfufv0tpyj#L829

**Comment by 1ucian0** [comment]
On hold until a discussion in #730 reaches consensus.

**Comment by jaygambetta** [comment]
have we decided what we are going to do about this?

**Comment by delapuente** [comment]
No longer on hold, let's move forward. @mtreinish do you mind to solve the conflicts with `master`?

**Comment by mtreinish** [comment]
The xunit xml output is blocked by a packaging bug on windows in the upstream project that converts the subunit v2 format output stream from stestr into the xunit xml format. I've pushed a PR here: https://github.com/testing-cabal/subunit/pull/33 to try and fix it. Once that merges and is in a release we can add that back. Alternatively we can carry a small (like 3-5 line) script somewhere in the repo to perform the conversion. If you think it's worth doing I can add a patch for the script and call that in appveyor.

File: appveyor.yml [hunk 1]
```diff
@@ -22,14 +22,18 @@ install:
...
 test_script:
-    - py.test -v --tb=short --junit-xml=test_results.xml
+    - stestr run
+    - stestr last --subunit > test_results.subunit
 
 # Upload the test results to appveyor so they are shown on the "Tests" tab.
 on_finish:
     - ps: $wc = New-Object 'System.Net.WebClient'
-    - ps: $wc.UploadFile("https://ci.appveyor.com/api/testresults/junit/$($env:APPVEYOR_JOB_ID)", (Resolve-Path .\test_results.xml))
+# TODO(mtreinish): uncomment this when appveyor doesn't error on upload
```
Comment (delapuente) [inline appveyor.yml]: Could you link the issue in testing-cabal/subunit#33 with this `TODO`?
```suggestion
# TODO(mtreinish): uncomment this when testing-cabal/subunit#33 is fixed
```

Comment (mtreinish) [reply appveyor.yml]: Sure, will do

**Pull request merged**